lib/blocks.php

```diff
@@ -45,3 +45,20 @@ function gutenberg_reregister_core_block_types() {
 	}
 }
 add_action( 'init', 'gutenberg_reregister_core_block_types' );
+
+/**
+ * Registers a new block style.
+ *
+ * @param string $block_name       Block type name including namespace.
+ * @param array  $style_properties Array containing the properties of the style name, label, style (name of the stylesheet to be enqueued), inline_style (string containing the CSS to be added).
+ */
+function register_block_style( $block_name, $style_properties ) {
```

Comment:
How can someone unregister a block style using PHP, when it was registered using PHP?

Reply:
Good question @swisspidy. We have three options to implement unregister functionality:
1 - A unregister_block_style function that uses the global wp_filters to get the class instance used in the filter and then calls remove_action. (seems hacky but should work well).
2 - Register block style stores the WP_Block_Styles_Enqueue_Block_Assets_Action in a global array. unregister_block_style function retrieves the instance from the global array using (block name and style name) and calls remove_action.
3 - Plugin developers pass an instance of WP_Block_Styles_Enqueue_Block_Assets_Action (maybe we should find a better name in that case) to register_block_style and they can save the instance to later call unregister_block_style. This does not allow another plugin to remove a style added as they may not have access to the instance.

It seems using the current register API option 2 is the best, but using a global variable is not ideal.

Another possibility is using a new class BlockStyles with a static method to register and another to unregister. The class could follow the same approach specific in 2 but using an instance variable instead of a global variable.


Any ideas about what approach to follow? Is there an alternative approach that I'm missing and is better than the ones referred?

Reply:
> Another possibility is using a new class BlockStyles with a static method to register and another to unregister. The class could follow the same approach specific in 2 but using an instance variable instead of a global variable.

👍 

This reminds me a lot of `WP_Block_Type_Registry`, which basically does the same thing.

----

Now that I think about it...

Perhaps – since this is about adding some information for an existing block type - we could even re-use `WP_Block_Type_Registry` here. Example (off the top of my head):

```php
function register_block_style( $block_name, $style_properties ) {
	$block = null;

	if ( WP_Block_Type_Registry::is_registered( $block_name ) ) {
		$block = WP_Block_Type_Registry::unregister( $block_name );
	} else {
		$block = new WP_Block_Type( $block_name );
	}

	// This property would need to be added to WP_Block_Type in core eventually.
	$block->style_properties = isset( $block->style_properties ) ? $block->style_properties : array();
	$block->style_properties[] = $style_properties;

	WP_Block_Type_Registry::register( $block_name );
}

// This could be added directly to wp_enqueue_registered_block_scripts_and_styles()
function gutenberg_enqueue_additional_styles() {
	global $current_screen;

	$is_editor = ( ( $current_screen instanceof WP_Screen ) && $current_screen->is_block_editor() );

	$block_registry = WP_Block_Type_Registry::get_instance();
	foreach ( $block_registry->get_all_registered() as $block_name => $block_type ) {
		// Front-end styles.
		if ( $is_editor && ! empty( $block_type->style_properties ) ) {
			foreach( $block_type->style_properties as $properties ) {
				// Do all the sanity checks etc.
				wp_add_inline_script( /* ... */ );
			}
		}
	}
}

add_action( 'enqueue_block_assets', 'gutenberg_enqueue_additional_styles' );
```

Just thinking out loud here 🙂 Perhaps some other voices from core PHP devs would be helpful here, perhaps from @pento.

Reply:
I'm not sure if we should reuse WP_Block_Type_Registry as styles and blocks are not exactly the same in the client we use different structures. I agree the problem is the same. In php, it should also be possible to add styles to blocks that the server is not even aware they exist (registered with JS) I will probably use a WP_Block_Styles_Registry as a registry for styles, and follow path 2 instead of using a global I will use the styles registry.